internal/ad/backends/winbind/winbind_test.go

```diff
@@ -0,0 +1,187 @@
+package winbind_test
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/adsys/internal/ad/backends/winbind"
+	"github.com/ubuntu/adsys/internal/testutils"
+)
+
+func TestWinbind(t *testing.T) {
+	// Build mock libwbclient
+	var mockLibPath string
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		mockLibPath = buildMockLib(t)
+	}
+
+	// We setup and rerun in a subprocess because we need to preload the mock libwbclient
+	if setupSubprocessForTest(t, mockLibPath) {
+		return
+	}
+
+	tests := map[string]struct {
+		wbclientBehavior string
+		staticADDomain   string
+		staticADServer   string
+
+		wantErr bool
+	}{
+		"Lookup is successful": {},
+
+		// Override cases
+		"Lookup with overridden ad_domain":                  {staticADDomain: "overridden.com"},
+		"Lookup with overridden ad_server":                  {staticADServer: "controller.overridden.com"},
+		"Lookup with overridden ad_server with LDAP prefix": {staticADServer: "ldap://controller.overridden.com"},
+
+		// Error cases
+		"Error looking up domain":      {wbclientBehavior: "domain_not_found", wantErr: true},
+		"Error looking up DC name":     {wbclientBehavior: "error_getting_dc_name"},
+		"Error getting online status":  {wbclientBehavior: "error_getting_online_status"},
+		"Error when domain is offline": {wbclientBehavior: "domain_is_offline"},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			// Set up mock libwbclient behavior
+			t.Setenv("ADSYS_WBCLIENT_BEHAVIOR", tc.wbclientBehavior)
+
+			config := winbind.Config{}
+			if tc.staticADDomain != "" {
+				config.ADDomain = tc.staticADDomain
+			}
+			if tc.staticADServer != "" {
+				config.ADServer = tc.staticADServer
+			}
+
+			winbind, err := winbind.New(context.Background(), config)
+			if tc.wantErr {
+				require.Error(t, err, "New should have errored out")
+				return
+			}
+
+			var got bytes.Buffer
+			got.WriteString(fmt.Sprintf("* Domain(): %s\n", winbind.Domain()))
+
+			serverURL, err := winbind.ServerURL(context.Background())
+			serverLine := fmt.Sprintf("* ServerURL(): %s\n", serverURL)
+			if err != nil {
+				serverLine = fmt.Sprintf("* ServerURL ERROR(): %s\n", err)
+			}
+			got.WriteString(serverLine)
+
+			isOnline, err := winbind.IsOnline()
+			isOnlineLine := fmt.Sprintf("* IsOnline(): %t\n", isOnline)
+			if err != nil {
+				isOnlineLine = fmt.Sprintf("* IsOnline ERROR(): %s\n", err)
+			}
+			got.WriteString(isOnlineLine)
+
+			got.WriteString(fmt.Sprintf("* HostKrb5CCNAME(): %s\n", winbind.HostKrb5CCNAME()))
+			got.WriteString(fmt.Sprintf("* DefaultDomainSuffix(): %s\n", winbind.DefaultDomainSuffix()))
+			got.WriteString(fmt.Sprintf("* Config():\n%s\n", winbind.Config()))
```

Comment:
Thid is a lot of repetition between the 2 backends. Maybe we can just create a backend test helper package, that are importing only by those 2 test files, which takes a writer + the result of all the calls and write/format it for us?

Reply:
Excellent! I was almost in favor of doing the golden file update/comparison in my initial comment, but then, I think this is less flexible in the long term, so let’s keep it as you did, apart (chaining small helpers together).